src/lib.rs

```diff
@@ -137,6 +137,46 @@ impl<T: EventSubscriber + ?Sized> MutEventSubscriber for T {
     }
 }
 
+impl MutEventSubscriber for Box<dyn MutEventSubscriber> {
```

Comment:
Just wondering, isn't it enough to only add

```rust
impl<T: MutEventSubscriber + ?Sized> MutEventSubscriber for Box<T> {
    fn process(&mut self, events: Events, ops: &mut EventOps) {
        self.deref_mut().process(events, ops);
    }

    fn init(&mut self, ops: &mut EventOps) {
        self.deref_mut().init(ops);
    }
}
```

and specify the `Send` and/or `Sync` trait bound when declaring the `EventManager` object (i.e. something like `let m = EventManager<Box<&dyn MutEventSubscriber + Sync + Send>> ...`?

This also makes things work for non-dyn `Box<T: MutEventSubscriber>`s, which I forgot to add previously :(

Reply:
That's what I was trying to hint at with my commment from the previous PR (really poorly though :)) ).

In the meantime, I cherry-picked these commits on top of my integration tests branch, and wrote a dummy test that works for `Box` as well. @jiangliu am I missing something?

Here is the patch on top of you patches & my integration-tests branch:
```
diff --git a/src/lib.rs b/src/lib.rs
index 6c5ee3a..dbcd81e 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -5,7 +5,7 @@
 #![deny(missing_docs)]
 
 use std::cell::RefCell;
-use std::ops::{Deref, DerefMut};
+use std::ops::Deref;
 use std::rc::Rc;
 use std::result;
 use std::sync::{Arc, Mutex};
@@ -140,43 +140,13 @@ impl<T: EventSubscriber + ?Sized> MutEventSubscriber for T {
     }
 }
 
-impl MutEventSubscriber for Box<dyn MutEventSubscriber> {
-    fn process(&mut self, events: Events, ops: &mut EventOps) {
-        self.deref_mut().process(events, ops);
-    }
-
-    fn init(&mut self, ops: &mut EventOps) {
-        self.deref_mut().init(ops);
-    }
-}
-
-impl MutEventSubscriber for Box<dyn MutEventSubscriber + Send> {
-    fn process(&mut self, events: Events, ops: &mut EventOps) {
-        self.deref_mut().process(events, ops);
-    }
-
-    fn init(&mut self, ops: &mut EventOps) {
-        self.deref_mut().init(ops);
-    }
-}
-
-impl MutEventSubscriber for Box<dyn MutEventSubscriber + Sync> {
-    fn process(&mut self, events: Events, ops: &mut EventOps) {
-        self.deref_mut().process(events, ops);
-    }
-
-    fn init(&mut self, ops: &mut EventOps) {
-        self.deref_mut().init(ops);
-    }
-}
-
-impl MutEventSubscriber for Box<dyn MutEventSubscriber + Send + Sync> {
-    fn process(&mut self, events: Events, ops: &mut EventOps) {
-        self.deref_mut().process(events, ops);
+impl<T: EventSubscriber + ?Sized> EventSubscriber for Box<T> {
+    fn process(&self, events: Events, ops: &mut EventOps) {
+        self.deref().process(events, ops);
     }
 
-    fn init(&mut self, ops: &mut EventOps) {
-        self.deref_mut().init(ops);
+    fn init(&self, ops: &mut EventOps) {
+        self.deref().init(ops);
     }
 }
 
diff --git a/tests/multi_threaded.rs b/tests/multi_threaded.rs
index 31e1c8a..c242836 100644
--- a/tests/multi_threaded.rs
+++ b/tests/multi_threaded.rs
@@ -144,3 +144,28 @@ fn test_subsscriber_inner_mut() {
 
     assert_eq!(subscriber.counter(), 100);
 }
+
+#[test]
+fn test_subsscriber_inner_mut_2() {
+    let mut event_manager = EventManager::<Box<dyn EventSubscriber + Send + Sync>>::new()
+        .expect("Cannot create event manager");
+
+    let subscriber = Box::new(CounterInnerMutSubscriber::new());
+    subscriber.trigger_event();
+
+    // Let's just clone the subscriber before adding it to the event manager. This will allow us
+    // to use it from this thread without the need to call into EventManager::subscriber_mut().
+    let subscriber_id = event_manager.add_subscriber(subscriber);
+
+    let thread_handle = thread::spawn(move || {
+        for _ in 0..100 {
+            // When the event manager loop exits, it will return the number of events it
+            // handled.
+            let ev_count = event_manager.run().unwrap();
+            assert_eq!(ev_count, 1);
+        }
+
+        assert!(event_manager.remove_subscriber(subscriber_id).is_ok());
+    });
+    thread_handle.join().unwrap();
+}
\ No newline at end of file

```

Reply:
> ```rust
> impl<T: MutEventSubscriber + ?Sized> MutEventSubscriber for Box<T> 
> ```

This fails with:
error[E0119]: conflicting implementations of trait `MutEventSubscriber` for type `std::boxed::Box<_>`:
   --> libs/event-manager/src/lib.rs:143:1
    |
133 | impl<T: EventSubscriber + ?Sized> MutEventSubscriber for T {
    | ---------------------------------------------------------- first implementation here
...
143 | impl<T: MutEventSubscriber + ?Sized> MutEventSubscriber for Box<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::boxed::Box<_>`
    |
    = note: downstream crates may implement trait `EventSubscriber` for type `std::boxed::Box<_>`

Reply:
I tested it with this patch: https://github.com/andreeaflorescu/event-manager/commit/79586f89f18a3ce4481dfb064a838792eb307666 and it seems to be working.

You should implement it for `EventSubscriber` instead of `MutEventSubscriber`. The reason why it fails is that there is an implementation for `MutEventSubscriber` for `EventSubscriber`. When you add the Box implementation for mut, the 2 implementations will conflict.

Reply:
Fails in the same way:
error[E0277]: the trait bound `(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static): event_manager::EventSubscriber` is not satisfied
  --> src/utils/src/epoll_manager.rs:42:38
   |
42 |             mgr: Arc::new(Mutex::new(mgr)),
   |                                      ^^^ the trait `event_manager::EventSubscriber` is not implemented for `(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)`
   |
   = note: required because of the requirements on the impl of `event_manager::EventSubscriber` for `std::boxed::Box<(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)>`
   = note: required because of the requirements on the impl of `event_manager::MutEventSubscriber` for `std::boxed::Box<(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)>`
   = note: required by `event_manager::EventManager::<S>::new`

error[E0599]: no method named `add_subscriber` found for type `std::sync::MutexGuard<'_, event_manager::EventManager<std::boxed::Box<(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)>>>` in the current scope
  --> src/utils/src/epoll_manager.rs:52:28
   |
52 |                 return mgr.add_subscriber(handler);
   |                            ^^^^^^^^^^^^^^ method not found in `std::sync::MutexGuard<'_, event_manager::EventManager<std::boxed::Box<(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)>>>`
   |
   = note: the method `add_subscriber` exists but the following trait bounds were not satisfied:
           `event_manager::EventManager<std::boxed::Box<(dyn event_manager::MutEventSubscriber + std::marker::Send + 'static)>> : event_manager::SubscriberOps`

Reply:
Ok so the problem is created by the `impl<T: EventSubscriber + ?Sized> EventSubscriberMut for T` blanket implementation that was initially used to avoid writing too much boilerplate code. For things like `Box<T>`, which can be either `EventSubscriber + EventSubscriberMut`, or just `EventSubscriberMut` depending on `T`, it creates some ambiguity that makes Rust complain like Gerry noticed. Andreea's approach was working, bot only for `Box<dyn EventSubscriber + ...>`, and not for `Box<dyn MutEventSubscriber ...>`. I'll post a commit that proposes a fix soon. Sorry about that :(

Reply:
I posted a fix [here](https://github.com/alexandruag/event-manager/commit/835c9e76250f6a113a0cca567df5abbac69224e7). It might interfere with some of the examples/tests/doc tests in #26, which looks like it can be merged quite soon. Is it ok if we do that first, than I post the fix together with any required fixes, and then we return to this PR? Btw, I'm still thinking about the remote endpoint commit ... there some interesting questions raised by your proposal, and I hope to have answers by Monday :D

Reply:
Sure, I will repost once #26 has been merged.

Reply:
@alexandruag Could you please help to send out a MR for the refactoring?

Reply:
I've submitted #30 with the fix. The approach where you declare the manager along the lines of `let m = EventManager<Box<dyn MutEventSubscriber + Sync + Send>> ...` should work now. If not, please ping and I'll take another look ASAP.